avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -567,24 +605,33 @@ private void processUnion(JVar unionSchemaVar, final String name, final Schema u
       }
 
       if (action.getShouldRead()) {
-        // The reader's union could be re-ordered, so we need to find the one that matches.
-        for (int j = 0; j < readerUnionSchema.getTypes().size(); j++) {
-          Schema potentialReaderSchema = readerUnionSchema.getTypes().get(j);
-          // Avro allows unnamed types to appear only once in a union, but named types may appear multiple times and
-          // thus need to be disambiguated via their full-name (including aliases).
-          if (potentialReaderSchema.getType().equals(optionSchema.getType()) &&
-              (!schemaAssistant.isNamedType(potentialReaderSchema) ||
-                  AvroCompatibilityHelper.getSchemaFullName(potentialReaderSchema).equals(AvroCompatibilityHelper.getSchemaFullName(optionSchema)) ||
-                  potentialReaderSchema.getAliases().contains(AvroCompatibilityHelper.getSchemaFullName(optionSchema)))) {
-            readerOptionSchema = potentialReaderSchema;
-            readerOptionUnionBranchIndex = j;
-            break;
+        // If unionReaderSchema is not a union then only compatible writer union schema type can be processed,
+        // otherwise the readerOptionSchema will be set to null what should result in AvroTypeException as
+        // in vanilla avro.
+        if (readerSchemaNotAUnion) {
+            readerOptionSchema = (i == compatibleWriterSchema) ? unionReaderSchema : null;
+        } else {
+          // The reader's union could be re-ordered, so we need to find the one that matches.
+          for (int j = 0; j < unionReaderSchema.getTypes().size(); j++) {
+            Schema potentialReaderSchema = unionReaderSchema.getTypes().get(j);
+            // Avro allows unnamed types to appear only once in a union, but named types may appear multiple times and
```

Comment:
i think avro also allows some "conversions" - like between string and bytes. if im right (needs testing vs vanilla avro) you'd need to also look for these sorts of matches?